services/workflows-service/src/workflow/update-documents.ts

```diff
@@ -0,0 +1,77 @@
+import { DefaultContextSchema } from '@/workflow/schemas/context';
+import { PartialDeep } from 'type-fest';
+import { getDocumentId } from '@/workflow/utils';
+import { cloneDeep, merge } from 'lodash';
+
+type Documents = DefaultContextSchema['documents'];
+type Document = Documents[number];
+type Pages = Document['pages'];
+
+const toDocumentsMap = <T extends PartialDeep<Documents>>(documents: T) => {
+  return documents.reduce((map, document) => {
+    const documentId = getDocumentId(document as Document);
+
+    map[documentId] = cloneDeep(document);
+    delete map[documentId]?.pages;
+
+    return map;
+  }, {} as Record<string, T[number]>);
+};
+
+function updatePages(oldPages: Pages, newPages: Pages) {
+  const doesntHaveChanges = !newPages.length;
+
+  if (doesntHaveChanges) {
+    return Object.values(oldPages);
+  }
+
+  const hasNewPages = newPages.some(page => !page.ballerineFileId);
+
+  if (hasNewPages) {
+    return newPages;
+  }
+
+  return [
+    ...oldPages.map(oldPage => {
+      const newPage = newPages.find(newPage => newPage.ballerineFileId === oldPage.ballerineFileId);
+
+      return newPage ? merge(oldPage, newPage) : oldPage;
+    }),
+    ...newPages.filter(newPage => !newPage.ballerineFileId),
+  ];
+}
+
+const toDocumentPagesMap = (documents: Documents) => {
+  return documents.reduce((documentPagesMap, document) => {
+    const documentId = getDocumentId(document);
+
+    documentPagesMap[documentId] = cloneDeep(document.pages || []);
+
+    return documentPagesMap;
+  }, {} as Record<string, Pages>);
+};
+
+export function updateDocuments(
```

Comment:
To keep consistency with the rest of the code base let's use arrow functions where possible.